### ISSUE TITLE
Instrument buffers

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#4936]
We need to instrument buffer activity. The goals of buffer observability are:

- [ ] Buffers size in event quantity
- [ ] Buffer size in bytes
- [ ] Latency of the buffer (how long events are in the buffer)
- [ ] Buffer total/possible size (to determine % used)
- [ ] Events dropped quantity (if `when_full` is set to `drop_newest`)

I'm indifferent to the metrics names, as long as they are consistent with existing names.

**Comment by jszwedko** [comment]
I mentioned this to Will earlier, but I think we should punt on latency for now, as it is more involved to track, and just report the other metrics.

**Comment by binarylogic** [comment]
Agree, the metrics for the buffer should be very similar to the component metrics and be simple to start. I'll update the description.
